File: src/Comp/Assignment1/ProductionSystemSolver.java
package Comp.Assignment1;

import java.util.*;

/**
 * Created by Ali on 2017-01-19.
 */
public class ProductionSystemSolver {

    public static IProductionSystem BreadthFirstSearch(IProductionSystem productionSystem)
    {
        if(productionSystem.isSolved())
        {
            return productionSystem;
        }

        Queue<IProductionSystem> productionSystemQueue = new ArrayDeque<>();
        HashSet<Integer> visitedStates = new HashSet<>();

        productionSystemQueue.add(productionSystem);
        visitedStates.add(productionSystem.hashCode());

        while(!productionSystemQueue.isEmpty())
        {
            IProductionSystem currentState = productionSystemQueue.poll();

            for (IProductionSystem newState  : currentState.generateAllChildStates(currentState))
            {
                if(newState.isSolved())
                {
                   return newState;
                }

                if(!visitedStates.contains(newState.hashCode())) {
                    productionSystemQueue.add(newState);
                    visitedStates.add(newState.hashCode());
                }
            }
        }

        return null;
    }

    public static IProductionSystem DepthFirstSearch(IProductionSystem productionSystem)
    {
        if(productionSystem.isSolved())
        {
            return productionSystem;
        }

        Stack<IProductionSystem> productionSystemStack = new Stack<>();
        HashSet<Integer> visitedStates = new HashSet<>();

        productionSystemStack.push(productionSystem);
        visitedStates.add(productionSystem.hashCode());

        while(!productionSystemStack.isEmpty())
        {
            IProductionSystem currentState = productionSystemStack.pop();

            for (IProductionSystem newState : currentState.generateAllChildStates(currentState))
            {
                if(newState.isSolved())
                {
                    return newState;
                }

                if(!visitedStates.contains(newState.hashCode())) {
                    visitedStates.add(newState.hashCode());
                    productionSystemStack.push(newState);
                }
            }
        }

        return null;
    }

     public static IProductionSystem AStarSearch(IProductionSystem productionSystem) {

         if (productionSystem.isSolved()) {
             return productionSystem;
         }

         Comparator<IProductionSystem> comparator = new ProductionSystemComparator();
         Queue<IProductionSystem> open = new PriorityQueue<>(10, comparator);
         HashSet<IProductionSystem> closed = new HashSet<>();
            int i =0;
         open.add(productionSystem);

         while (!open.isEmpty()) {

             IProductionSystem currentState = open.poll();
             closed.add(currentState);

             for (IProductionSystem newState : currentState.generateAllChildStates(currentState))
             {
                 if (newState.isSolved()) {
                     return newState;
                 }

                 boolean inClosed = closed.contains(newState);
                 if(!inClosed)
                 {
                     boolean inOpen = open.contains(newState);
                     if(!inOpen)
                         open.add(newState);
                     else
                     {
                         IProductionSystem openState = getObject(open, newState);

                         if(newState.getStateValue() < openState.getStateValue())
                         {
                             openState.setStateValue(newState.getStateValue());
                             openState.setPreviousState(newState.getPreviousState());
                         }
                     }

                 }

             }
         }

         return null;
     }

    private static IProductionSystem reverseStates(IProductionSystem productionSystem)
    {
        IProductionSystem previousSystem = null;
        IProductionSystem currentSystem = productionSystem;

        while(currentSystem != null)
        {
            IProductionSystem temp =  currentSystem.getPreviousState();
            currentSystem.setPreviousState( previousSystem);
            previousSystem = currentSystem;
            currentSystem = temp;
        }

        return previousSystem;
    }


    public static String getPathToSuccess(IProductionSystem productionSystem)
    {
        StringBuilder builder = new StringBuilder();
        productionSystem = reverseStates(productionSystem);

        int count = 0;

        while(productionSystem != null)
        {
            builder.append(productionSystem.toString());
            productionSystem =  productionSystem.getPreviousState();
            count++;
        }

        return builder.toString() + "Length: " + count;
    }

     private static IProductionSystem getObject(
             Collection<IProductionSystem> productionSystems,
             IProductionSystem productionSystem)
     {
         for (IProductionSystem system: productionSystems) {
             if(productionSystem.equals(system))
                 return system;
         }

         return null;
     }
}

File: FinalProject/FinalProject/ProductionSystemSolver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FinalProject
{
    public static class ProductionSystemSolver
    {
        public static PlayerLineup BreadthFirstSearch(PlayerLineup productionSystem)
        {
            Queue<PlayerLineup> productionSystemQueue = new Queue<PlayerLineup>();
            HashSet<PlayerLineup> visitedStates = new HashSet<PlayerLineup>();

            productionSystemQueue.Enqueue(productionSystem);
            visitedStates.Add(productionSystem);

            while (productionSystemQueue.Count() > 0)
            {
                PlayerLineup currentState = productionSystemQueue.Dequeue();

                foreach(PlayerLineup newState in currentState.GenerateNextStates())
                {
                    if (!visitedStates.Contains(newState))
                    {
                        visitedStates.Add(newState);

                        if (newState.IsSolved())
                        {
                            return newState;
                        }
                        else 
                        {
                            productionSystemQueue.Enqueue(newState);
                        }
                    }
                }
            }

            return null;
        }

        public static PlayerLineup DepthFirstSearch(PlayerLineup productionSystem)
        {
            Stack<PlayerLineup> productionSystemQueue = new Stack<PlayerLineup>();
            HashSet<PlayerLineup> visitedStates = new HashSet<PlayerLineup>();

            productionSystemQueue.Push(productionSystem);
            visitedStates.Add(productionSystem);

            while (productionSystemQueue.Count() > 0)
            {
                PlayerLineup currentState = productionSystemQueue.Pop();

                foreach (PlayerLineup newState in currentState.GenerateNextStates())
                {
                    if (!visitedStates.Contains(newState))
                    {
                        visitedStates.Add(newState);

                        if (newState.IsSolved())
                        {
                            return newState;
                        }
                        else
                        {
                            productionSystemQueue.Push(newState);
                        }
                    }
                }
            }

            return null;
        }

        public static PlayerLineup AStarSearch(PlayerLineup productionSystem)
        {
            var open = new C5.IntervalHeap<PlayerLineup>(new PlayerLineupComparer());
            var closed = new HashSet<PlayerLineup>();
            open.Add(productionSystem);

            while (open.Count() > 0)
            {
                PlayerLineup currentState = open.FindMax();
                closed.Add(currentState);

                if (currentState.IsSolved())
                {
                    return currentState;
                }

                foreach (PlayerLineup newState in currentState.GenerateNextStates())
                {
                    var inClosed = closed.Contains(newState);

                    if (!inClosed)
                    {
                        var inOpen = open.Contains(newState);
                        if (!inOpen)
                            open.Add(newState);
                        else
                        {
                            PlayerLineup openState = open.First(line => line.Equals(newState));

                            if (newState.GetStateValue() > openState?.GetStateValue())
                            {
                                open.Add(newState);
                            }
                        }

                    }

                }
            }

            return null;
        }

        public static PlayerLineup HillClimbing(PlayerLineup productionSystem)
        {

            var neighbors = productionSystem.GenerateNextStates();
            var counter = 1;

            while(neighbors.Count() > 0)
            {
                PlayerLineup nextNode = null;
                double maxEval = Int32.MinValue;

                foreach(var node in neighbors)
                {
                    var currentValue = node.GetStateValue();
                    counter++;
                    if(currentValue > maxEval)
                    {
                        maxEval = currentValue;
                        nextNode = node;
                    }
                }

                if (nextNode.IsSolved())
                    return nextNode;
                else
                    neighbors = nextNode.GenerateNextStates();
            }
           
            return null;
        }
    }
}
